### PR TITLE
Fast dependency scanning for Swift

### PR DESCRIPTION
**Pull request by DougGregor**
Implement a new "fast" dependency scanning option,
`-scan-dependencies`, in the Swift frontend that determines all
of the source file and module dependencies for a given set of
Swift sources. It covers all forms of dependencies:

1) Swift (serialized) module files, by reading the module header
2) Swift interface files, by parsing the source code to find imports
3) Swift source modules, by parsing the source code to find imports
4) Clang modules, using Clang's fast dependency scanning tool
5) Bridging headers, including the source files and modules they depend on

A single `-scan-dependencies` operation maps out the full
dependency graph for the given Swift source files, including all
of the Swift and Clang modules that may need to be built, such
that all of the work can be scheduled up front by the Swift
driver or any other build system that understands this
option. The dependency graph is emitted as JSON, which can be
consumed by these other tools. Swift code with `Codable`
data structures is used for validating this JSON, and can
(eventually) be picked up by other tools.

This operation also supports `-emit-dependencies`, the flag that
produces make-style dependencies to determine all of the input
files that can affect the compilation, making it usable as a fast
dependency scanner even for clients not interesting in prebuilding
modules.

In a source file that imports just `Cocoa` on macOS, fast
dependency scanning with this patch is ~11x faster than
the current dependency scanner (with `-resolve-imports`), taking
it from 5.7s (and writing 48MB of Clang modules) down to .49s.

### CODE REVIEW

**Comment by DougGregor** [comment]
@Bigcheese this totally doesn't work yet, but it's getting closer to mapping out the various Clang + Swift module builds that are involved in a Swift compile. You can see my super hacky interaction with Clang's fast dependency scanner in `ClangModuleDependencyScanner.cpp`.

**Comment by DougGregor** [comment]
Hey, it works a little bit now with a Clang change to handle `@import` directives in Clang's dependency scanner.

**Comment by DougGregor** [comment]
The necessary Clang changes are at https://github.com/apple/llvm-project/pull/392

**Comment by slavapestov** [comment]
Would this allow the frontend to get out of the business of building interface files altogether?

**Comment by DougGregor** [comment]
> Would this allow the frontend to get out of the business of building interface files altogether?

Yes, that's the intent. The driver would coordinate the interface file builds so the frontend wouldn't have to do it.

**Comment by DougGregor** [comment]
I collapsed everything because it doesn't make sense for this to ride on top of `-emit-imported-modules`; rather, that flag should (separately) go away. Instead, implement a new flag (`-scan-dependencies`) whose primary output is the JSON dependency graph and which will eventually also emit normal makefile dependencies (`.d` file) as well, for clients that need it.

**Comment by jerrymarino** [comment]
So this should this allow us to directly run the frontend in external build systems directly? Seems like we can use the new .d files

**Comment by DougGregor** [comment]
> So this should this allow us to directly run the frontend in external build systems directly? Seems like we can use the new .d files

We'd prefer that external build systems go through the driver in general, because the frontend interface is  the supported way to call the compiler. However, we can expose this as a driver option once it's been tested a bit better. I would certainly appreciate if others would test it out and see what I've missed.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by DougGregor** [comment]
Brought it up to date with master. It no longer needs cherry-picks on top of Clang

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by davidungar** [comment]
Lovely, readable code. But, why were the storage classes factored out, instead of just having a base `ModuleDependencies` class and two subclasses? Is there not always a one-to-one relationship to the storage? I'm missing something here...

File: lib/ClangImporter/ClangModuleDependencyScanner.cpp [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+
+Optional<ModuleDependencies> ClangImporter::getModuleDependencies(
+    StringRef moduleName, ModuleDependenciesCache &cache) {
+  // Check whether there is already a cached result.
+  if (auto found = cache.findDependencies(
+          moduleName, ModuleDependenciesKind::Clang))
+    return found;
+
+  // Retrieve or create the shared state.
+  auto clangImpl = getOrCreateClangImpl(cache);
```
Comment (davidungar) [inline lib/ClangImporter/ClangModuleDependencyScanner.cpp:234]: This function is a bit long; Could this be a `const`?

File: lib/ClangImporter/ClangModuleDependencyScanner.cpp [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+      moduleName, ModuleDependenciesKind::Swift);
+
+  // If we've already recorded bridging header dependencies, we're done.
+  auto swiftDeps = targetModule.getAsSwiftModule();
+  if (!swiftDeps->bridgingSourceFiles.empty() ||
+      !swiftDeps->bridgingModuleDependencies.empty())
+    return false;
+
+  // Retrieve or create the shared state.
+  auto clangImpl = getOrCreateClangImpl(cache);
```
Comment (davidungar) [inline lib/ClangImporter/ClangModuleDependencyScanner.cpp:292]: `const`?

File: lib/ClangImporter/ClangModuleDependencyScanner.cpp [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    return None;
+  }
+
+  // Record module dependencies for each module we found.
+  recordModuleDependencies(cache, *clangDependencies);
+
+  return cache.findDependencies(moduleName, ModuleDependenciesKind::Clang);
+}
+
+bool ClangImporter::addBridgingHeaderDependencies(
```
Comment (davidungar) [inline lib/ClangImporter/ClangModuleDependencyScanner.cpp:279]: You know me; I would prefer to break up this function into more understandable (for me) pieces.

File: lib/ClangImporter/ClangModuleDependencyScanner.cpp [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+  // Retrieve or create the shared state.
+  auto clangImpl = getOrCreateClangImpl(cache);
+
+  // Reform the Clang importer options.
+  // FIXME: Just save a reference or copy so we can get this back.
+  ClangImporterOptions importerOpts;
+
+  // Retrieve the bridging header.
+  std::string bridgingHeader = *targetModule.getBridgingHeader();
```
Comment (davidungar) [inline lib/ClangImporter/ClangModuleDependencyScanner.cpp:299]: `const`?

File: lib/ClangImporter/ClangModuleDependencyScanner.cpp [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+
+  // Determine the command-line arguments for dependency scanning.
+  auto &ctx = Impl.SwiftContext;
+  std::vector<std::string> commandLineArgs =
```
Comment (davidungar) [inline lib/ClangImporter/ClangModuleDependencyScanner.cpp:303]: Sorry if all these  `consts` are getting tedious. As you know, the idea is that if there are more than a few lines after the assignment, if something is `const` the first assignment is enough to know where it's coming from without having to inspect all uses. (Esp. in C++ where one can pass a reference, sigh.) Forgive me for stating the obvious.

File: lib/Frontend/Frontend.cpp [hunk 6]
```diff
@@ -729,12 +729,11 @@ void CompilerInstance::addAdditionalInitialImportsTo(
 /// builds. This allows for use of popular specialized functions
 /// from the standard library, which makes the non-optimized builds
 /// execute much faster.
-static bool
-shouldImplicityImportSwiftOnoneSupportModule(CompilerInvocation &Invocation) {
-  if (Invocation.getImplicitModuleImportKind() !=
-      SourceFile::ImplicitModuleImportKind::Stdlib)
+bool CompilerInvocation::shouldImportSwiftONoneSupport() const {
```
Comment (davidungar) [inline lib/Frontend/Frontend.cpp:732]: Nice!

File: lib/FrontendTool/FrontendTool.cpp [hunk 7]
```diff
@@ -1255,8 +1256,10 @@ static bool performCompile(CompilerInstance &Instance,
     // Disable delayed parsing of type and function bodies when we've been
     // asked to dump the resulting AST.
     bool CanDelayBodies = Action != FrontendOptions::ActionType::DumpParse;
-    Instance.performParseOnly(/*EvaluateConditionals*/
-                    Action == FrontendOptions::ActionType::EmitImportedModules,
+    bool EvaluateConditionals =
+        Action == FrontendOptions::ActionType::EmitImportedModules
+        || Action == FrontendOptions::ActionType::ScanDependencies;
+    Instance.performParseOnly(EvaluateConditionals,
```
Comment (davidungar) [inline lib/FrontendTool/FrontendTool.cpp:1262]: The flow in here and through 1271 is confusing. The predicate here is `FrontendOptions::shouldActionOnlyParse(Action)` and I would expect a return right after line 1262.
But instead, it falls through to 1271: `if (Action == FrontendOptions::ActionType::Parse)` and it that is true it returns. Which leaves me pondering: what's the difference between the two predicates? Is one a subset of the other? There must be a clearer way to write this. Maybe also do a return right here?
OK, I see it, now. Will leave the above confused ramblings in FWIW.

The whole business starting at 1255 could use a comment to the effect of do the the right about of parsing, etc. for the mode, or some such. (I would factor it out into a function with that name.)

Then, a blank line after 1258 would make it clear that the comment above it only applies to that one line.
(Not your fault!)

Then, a "why" comment might be nice about why conditionals are only being evaluated in these two cases.

I guess what's going on is that this function could be split in a "parse" phase, through 1272 and an after-parse phase, and that demarcation would clarify what's going on. (Not your fault, but this change just makes it incrementally worse, like an amoeba that's overdue for fission.)

**Review by davidungar — APPROVED** [review]
I can't for correctness, but overall it looks pretty good. I threw in a few suggestions that could incrementally improve the clarity, IMO. The most important of those would be the places where a "why" comment could help.

File: include/swift/AST/ModuleDependencies.h [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines data structures for capturing module dependencies.
+//
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply include/swift/AST/ModuleDependencies.h:14]: The first part, sure, but this part of the code isn't the "fast" bit---it's just a description of the actual module dependencies.

File: include/swift/AST/ModuleDependencies.h [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+
+  virtual ModuleDependenciesStorageBase *clone() const = 0;
+
+  virtual ~ModuleDependenciesStorageBase();
+
+  /// The path to the compiled module file.
+  const std::string compiledModulePath;
+
+  /// The set of modules on which this module depends.
+  std::vector<std::string> moduleDependencies;
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply include/swift/AST/ModuleDependencies.h:56]: It describes "what has to be built before me". I'll clarify this in the overall file/class comment.

File: lib/AST/ASTContext.cpp [hunk 10]
```diff
@@ -1449,6 +1450,21 @@ void ASTContext::addModuleLoader(std::unique_ptr<ModuleLoader> loader,
   getImpl().ModuleLoaders.push_back(std::move(loader));
 }
 
+Optional<ModuleDependencies> ASTContext::getModuleDependencies(
+    StringRef moduleName, bool isUnderlyingClangModule,
+    ModuleDependenciesCache &cache) {
+  for (auto &loader : getImpl().ModuleLoaders) {
+    if (isUnderlyingClangModule &&
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply lib/AST/ASTContext.cpp:1457]: This code stinks; it'll be removed when we refactor more.

File: lib/AST/ModuleDependencies.cpp [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+  switch (kind) {
+  case ModuleDependenciesKind::Swift:
+    return SwiftModuleDependencies;
+
+  case ModuleDependenciesKind::Clang:
+    return ClangModuleDependencies;
+  }
+}
+
+bool ModuleDependenciesCache::hasDependencies(
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply lib/AST/ModuleDependencies.cpp:122]: Redundant, IMO.

File: lib/ClangImporter/ClangModuleDependencyScanner.cpp [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+  llvm::ErrorOr<StringRef> getImportHackFile();
+};
+
+ClangModuleDependenciesCacheImpl::~ClangModuleDependenciesCacheImpl() {
+  if (!importHackFile.empty()) {
+    llvm::sys::fs::remove(importHackFile);
+  }
+}
+
+llvm::ErrorOr<StringRef> ClangModuleDependenciesCacheImpl::getImportHackFile() {
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply lib/ClangImporter/ClangModuleDependencyScanner.cpp:62]: Clang will provide us with an API that makes this unnecessary. It's a HACK.

**Comment by DougGregor** [comment]
Rebasing to fix merge conflicts, adding some comments, tweaking tests... and let's get this landed.

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test platform Windows

**Comment by DougGregor** [comment]
@swift-ci please test platform Windows

**Comment by DougGregor** [comment]
@swift-ci please test

**Pull request merged**